src/aspire/source/__init__.py

```diff
@@ -370,6 +373,151 @@ def whiten(self, noise_filter):
+            noise = (images * np.expand_dims(noise_mask, 2))
+            signal_mean += np.sum(signal)
+            noise_mean += np.sum(noise)
+        signal_denominator = self.n * np.sum(signal_mask)
+        noise_denominator = self.n * np.sum(noise_mask)
+        signal_mean /= signal_denominator
+        noise_mean /= noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to reverse contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to reverse contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, bg_radius=1.0, single_img=True, do_ramp=True,
+                             batch_size=512):
+        """
+        Normalize the images by the noise background
+
+        This is done by shifting the image density by the mean value of background
+        and scaling the image density by the standard deviation of background.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending the `Add` and `Multiple` filters to the generation pipeline.
+
+        :param bg_radius: Radius cutoff to be considered as background (in image size)
+        :param single_img: When it is `True`, normalize background based on the mean
+            and std values from corresponding individual images.
+        :param do_ramp: When it is `True`, fit a ramping background to the data
+            and subtract. Namely perform normalization based on values from each image.
+            Otherwise, a constant background level from all images is used.
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info(f'Normalize background on source object with radius size of {bg_radius}')
+        L = self.L
+        grid = grid_2d(L)
+        mask = (grid['r'] > bg_radius)
+
+        if do_ramp:
+            # Create matrices and reshape the background mask
+            # for fitting a ramping background
+            ramp_mask = np.vstack((grid['x'][mask].flatten(),
+                                   grid['y'][mask].flatten(),
+                                   np.ones(grid['y'][mask].flatten().size))).T
+            ramp_all = np.vstack((grid['x'].flatten(), grid['y'].flatten(),
+                                  np.ones(L*L))).T
+            mask_reshape = mask.reshape((L*L))
+
+        if single_img:
+            first_moment = np.zeros(self.n)
+            second_moment = np.zeros(self.n)
+            for i in range(0, self.n, batch_size):
+                images = self.images(start=i, num=batch_size).asnumpy()
+                if do_ramp:
```

Comment:
This doesn't reproduce the results of `im_normalize_bg` in the Matlab code. In `im_normalize_bg`, we subtract the fitted ramp, then calculate the mean and variance which normalize the *ramp-subtracted* images. Here, the ramp is fitted and subtracted from the images and those images are used to calculate first- and second-order moments, then they are thrown away. The pipeline later applies the normalization using mean and variance to the *original* images, not the ramp-subtracted ones.

The are two possible solutions:

- We create a new Xform to perform the ramp subtraction given the fitted ramp coefficients.
- If we discard the `single_img == False` option (as suggested above), the whole normalization can be done in the pipeline, since it works on a per-image basis. In that case, we simply add some sort of `LambdaXform` (does that exist?) which will call some variant of `im_normalize_bg` on the given images by fitting the ramp, subtracting, and normalizing by mean and variance.

The first option is a little cumbersome, but has the advantage of working with `single_img == False`. I think the simplest approach is to go with `single_img == True` always and calculate the normalization in the pipeline as in the second option.

Reply:
I probably will remove the Single==False. If so, how often we should calculate the ramp fitting, each image or each batch?

Reply:
Each image since we would have `single_img == True` all the time.

Reply:
Yes. I removed the `single_img==False` and reimplemented the do_ramp preprocess as pipeline in the new commits.